src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs

```diff
@@ -138,14 +138,18 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
             // order. We just need to iterate them and use the first one that can generate a link.
             var matches = _linkGenerationTree.GetMatches(context);
 
-            foreach (var match in matches)
+            // Perf: Avoid allocation of enumerator if matches is empty
+            if (matches?.Count > 0)
```

Comment:
Previously it used to use an enumerator . I think this change is not needed once I moved it to indexer